MoltenVK/MoltenVK/GPUObjects/MVKSwapchain.mm

```diff
@@ -408,15 +415,43 @@ static inline CIE1931XY VkXYColorEXTToCIE1931XY(VkXYColorEXT xy) {
 VkResult MVKSwapchain::getRefreshCycleDuration(VkRefreshCycleDurationGOOGLE *pRefreshCycleDuration) {
 	if (_device->getConfigurationResult() != VK_SUCCESS) { return _device->getConfigurationResult(); }
 
+#if MVK_IOS_OR_TVOS || MVK_MACCAT
 	NSInteger framesPerSecond = 60;
-#if MVK_IOS_OR_TVOS
 	UIScreen* screen = [UIScreen mainScreen];
 	if ([screen respondsToSelector: @selector(maximumFramesPerSecond)]) {
 		framesPerSecond = screen.maximumFramesPerSecond;
 	}
 #endif
-#if MVK_MACOS
-	// TODO: hook this up for macOS, probably need to use CGDisplayModeGetRefeshRate
+
+#if MVK_MACOS && !MVK_MACCAT
+	// Find the screen for the window whose content view is backed by the swapchains CAMetalLayer.
+	// Default to the mainScreen if no such window can be found.
+	NSScreen* screen = [NSScreen mainScreen];
+	CALayer* layer = _mtlLayer;
+	while (layer.superlayer) {
+		layer = layer.superlayer;
+	}
+
+	for (NSWindow* window in [[NSApplication sharedApplication] windows]) {
+		NSView *view = [window contentView];
+		if (view && ([view layer] == layer)) { // Check against layer and not _mtlLayer.
+			screen = [window screen];
+		}
+	}
+
+#if MVK_XCODE_13
+    // macOS 12+ only
+    double framesPerSecond = [screen maximumFramesPerSecond];
```

Comment:
I believe this also needs to be guarded with a
``	if ([screen respondsToSelector: @selector(maximumFramesPerSecond)]) {``
like it is done in the iOS path. the MVK_XCODE_13 would protect at compile time, but if the code successfully compiled on macOS 12+ would execute on an earlier macOS version, this would fail?

Also, the maximumFramesPerSecond assignment should only be used as a fallback if the existing CGDisplayModeGetRefreshRate path returns zero, not as a replacement for it on macOS 12+, because the maximumFramesPerSecond assignment is inferior, only minimally better than hard-coding 60 Hz:

1. We don't know if this gives valid results at all on standard fixed refresh rate displays - the method is for VRR adaptive sync displays. Essentially for those built-in pro-motion displays.
2. vkGetRefreshCycleDuration... is supposed to report the current framerate of a display, not the theoretical maximum rate, which may or may not be equal to the current rate at any given query time.

I think what you'd want is

1. CGDisplayModeGetRefreshRate
2. If that fails and MVK_XCODE_13 and respondsToSelector maximumFramesPerSecond, try that.
3. If that fails/is unsupported/gives a zero result, use hard-coded 60 Hz assignment.

Or for maximum correctness, CGDisplayModeGetRefreshRate and if that fails, the mentioned complex low-level method that glfw uses here https://github.com/glfw/glfw/blob/62e175ef9fae75335575964c845a302447c012c7/src/cocoa_monitor.m#L223

Reply:
Agreed. 

It's complicated because legacy support is messy...but both `#if MVK_XCODE_13` and `respondsToSelector:` are required. And this should just be moved down to the fallback area below, with something like:

```
if (framesPerSecond == 0) {
    framesPerSecond = 60.0;
#if MVK_XCODE_13
    if ([screen respondsToSelector: @selector(maximumFramesPerSecond)]) {
        framesPerSecond = [screen maximumFramesPerSecond];
    }
#endif
}
```

Reply:
True, forgot that this is supposed to be used as a fallback. And I really don't think it's really worth it implementing the complex GLFW code as it wouldn't offer too much, as otherwise we've already covered every possibility without using complex, and weird code. 

From NSScreen's documentation, there doesn't seem to be a way to find out what the current refresh rate is in regards to VRR capability and we'll just need to stick with the maximumRefreshRate, as there's not going to be many use cases where you'd want to render something with Vulkan/Metal and use a lower refresh rate. I think the only other way would be to actually measure the time between presents, given that the GPU isn't fully maxed and the presents are only limited in time by the presentation engine.